e2e: Trim junit reporter to adapt with testgrid

chendave opened the pull request:
The PR trimmed the system out and system err from the junit xml file, besides that, the concrete steps executed will be dropped as well if the testcase is not failed.  This makes the final report mostly like a V1, no system error and only show system out when the test failed, and no concrete steps printed in the file.

The trimmed version could be used for the tool like testgrid, ~the original one will still available under the same directory.~

/kind bug

Signed-off-by: Dave Chen <dave.chen@arm.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/kubernetes/issues/111510

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```


/sig testing

chendave commented on test/e2e/e2e_test.go:
```diff
@@ -181,4 +185,25 @@ var _ = ginkgo.ReportAfterSuite("Kubernetes e2e suite report", func(report ginkg
+// trimmedJunitReporter remove system error and system out element from the junit file which help the tool like testgrid to display the test status correctly.
+// please see: https://github.com/kubernetes/kubernetes/issues/111510
+func trimmedJunitReporter(report ginkgo.Report) {
+	reports := report.SpecReports
```
this is shall copy, I feel we need to deep copy to keep two reporter here.

chendave commented:
I test this locally, it will trim 40% space, let's check it by those jobs.


/cc @aojea @pohly

aojea commented:
I don't think we want to supress all the messages, this tool considers a specific size per example

https://github.com/kubernetes/kubernetes/pull/109112

The way I see it is that:
1. we should have a file with all the data, Onsi commented several times we can output all the data as json and store it in the job.
2. we can not have junit files larger than 200 mb (I think this is the testgrid limitation). Or either we rotate or we trim, right?

chendave commented:
> we should have a file with all the data, Onsi commented several times we can output all the data as json and store it in the job.

updated a code little bit, now we will have two files, 
``` junit_01_trimmed.xml junit_01.xml```

`junit_01.xml` contains all original information, and `junit_01_trimmed.xml` trimmed all those info added in V2, I compared with V1 reporter, the trimmed version should be similar with V1.


> we can not have junit files larger than 200 mb (I think this is the testgrid limitation). Or either we rotate or we trim, right?

I am not sure the size of the trimmed version, since I have no way to trigger the job and check it.

chendave commented:
> I don't think we want to supress all the messages, this https://github.com/kubernetes/kubernetes/pull/109112 considers a specific size per example

Those info are added in V2, and it is not consumed by the testgrid.  

Something I am worried is that whether Spyglass or testgrid would be able to load the right content from two different files.

chendave commented:
@pohly @aojea  should you have any other solutions, pls feel free to supersede this one.

aojea commented:
> Those info are added in V2, and it is not consumed by the testgrid.

then forget my comment, I misunderstood it

pohly commented:
Do we really need the non-trimmed version? Who is going to use it?

We also need to consider how we enable writing of the JUnit file. We currently have some extra code to ensure that `-report-dir` enables writing of the "normal" JUnit file. Perhaps we should make it so that `-report-dir` enables writing of the trimmed JUnit file and `ginkgo --junit-report` enables the "normal" one? We don't use `--junit-report` anywhere in Kubernetes because it is new and it is off by default, so only users who explicitly want it have to pay the price of writing the larger file.

chendave commented:
> Do we really need the non-trimmed version?

I am open to this, let's get more feedback from others @aojea @BenTheElder @dims  WDYT?

chendave commented:
Add some background for v1 vs. v2.

- v2 add an element `system-err` for all the testcases,  even the testcases passed or skipped.

e.g. 

```xml
          <testcase name="[It] [sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Generic Ephemeral-volume (block volmode) (late-binding)] ephemeral should support expansion of pvcs created for ephemeral pvcs" classname="Kubernetes e2e suite" status="skipped" time="0">
              <skipped message="skipped"></skipped>
              <system-err>[ReportAfterEach] TOP-LEVEL&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:142&#xA;</system-err>
          </testcase>
```


- when a testcase failed, v2 has `system-out` to tack the steps that were executed, and the information printed to stdout, and `system-err` has more information on the failure, e.g. stacktrace.

```xml
<system-err>[BeforeEach] [sig-network] Services&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:186&#xA;�      [1mSTEP:�[0m Creating a kubernetes client �[38;5;243m07/15/22 16:35:44.485�[0m&#xA;Jul ... dave:create an error anyway - conformance &lt;nil&gt;&#34;,&#xA;    }&#xA;Jul 15 16:35:44.528: FAIL: failed to fetch the service object for the service named kubernetes: dave:create an error anyway - conformance       &lt;nil&gt;&#xA;&#xA;Full Stack Trace&#xA;k8s.io/kubernetes/test/e2e/network.glob..fun      c25.3()&#xA;&#x9;/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/
```

- when a testcase is succeeded, the `system-out` and `system-err`  are also filled with the info of execution steps and logs. 
e.g.

```xml
              <system-out>&#xA;Report Entries:&#xA;By Step&#xA;/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:200&#xA;2022-08-02T15:32:45.542356862+08:00&#xA;&amp;{Text:Creating a kubernetes client Duration:0s}&#xA;--&#xA;By Step&#xA;/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:240&#xA;2022-08-02T15:32:45.544931733+08:00&#xA;&amp;{Text:Building a namespace api object, basename services Duration:0s}&#xA;--&#xA;By Step&#xA;/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:249&#xA;2022-08-02T15:32:45.563940578+08:00&#xA;&amp;{Text:Waiting for a default service account to be provisioned in namespace Duration:0s}&#xA;--&#xA;By Step&#xA;/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:252&#xA;2022-08-02T15:32:45.566584261+08:00&#xA;&amp;{Text:Waiting for kube-root-ca.crt to be provisioned in namespace Duration:0s}&#xA;--&#xA;By Step&#xA;/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:395&#xA;2022-08-02T15:32:45.575970729+08:00&#xA;&amp;{Text:Destroying namespace &#34;services-6094&#34; for this suite. Duration:0s}&#xA;</system-out>
              <system-err>[BeforeEach] [sig-network] Services&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:186&#xA;�[1mSTEP:�[0m Creating a kubernetes client �[38;5;243m08/02/22 15:32:45.542�[0m&#xA;Aug  2 15:32:45.542: INFO: &gt;&gt;&gt; kubeConfig: /etc/kubernetes/admin.conf&#xA;�[1mSTEP:�[0m Building a namespace api object, basename services �[38;5;243m08/02/22 15:32:45.544�[0m&#xA;�[1mSTEP:�[0m Waiting for a default service account to be provisioned in namespace �[38;5;243m08/02/22 15:32:45.563�[0m&#xA;�[1mSTEP:�[0m Waiting for kube-root-ca.crt to be provisioned in namespace �[38;5;243m08/02/22 15:32:45.566�[0m&#xA;[BeforeEach] [sig-network] Services&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service.go:758&#xA;[It] should provide secure master service  [Conformance]&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service.go:781&#xA;[AfterEach] [sig-network] Services&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:187&#xA;Aug  2 15:32:45.573: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready&#xA;�[1mSTEP:�[0m Destroying namespace &#34;services-6094&#34; for this suite. �[38;5;243m08/02/22 15:32:45.575�[0m&#xA;[AfterEach] [sig-network] Services&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service.go:762&#xA;[ReportAfterEach] TOP-LEVEL&#xA;  /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:145&#xA;</system-err>
```

pohly commented:
I've been checking where the JUnit "stdout" entries like this come from:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/111627/pull-kubernetes-e2e-kind-ipv6/1554370753451790336
```
Report Entries:
By Step
test/e2e/framework/framework.go:200
2022-08-02T07:59:02.423254009Z
&{Text:Creating a kubernetes client Duration:0s}
--
...
```
 
It turns out that these are the result of the ginkgo.By calls. In the past, those only emitted pretty-printed text output. The escape codes only work in a console, not in Spyglass, which already wasn't very nice.


Now we also get this useless entries in the JUnit XML.

We can deal with this in two ways:
- trim ReportEntries with `By Step` as name (could go into this PR)
- replace all `ginkgo.By` calls with normal framework.Logf calls (would also get rid of the annoying (?) escape sequences

I think we should do the trimming (can be done easily without much code churn and then consider whether we really want `ginkgo.By`.

pohly commented:
> when a testcase failed, v2 has system-out to tack the steps that were executed, and the information printed to stdout,

Do you have an example where `system-out` contains anything other than the "Report Entries", i.e. some actual information printed to stdout?

pohly commented:
[`ginkgo.By`](https://github.com/onsi/ginkgo/blob/v2.1.4/core_dsl.go#L496-L517) becomes more useful when used like this:

```
ginkgo.By("Setup", func() {
   // Setup code here
})
ginkgo.By("Test", func() {
   // Do something.
})
ginkgo.By("Verification", func() {
  // Check results.
})
```

The the duration of the function call will get included in the JUnit report entry. But that is not how we use it in Kubernetes...

chendave commented:
> Do you have an example where system-out contains anything other than the "Report Entries"

I don't have one atm, here is code comments on the `CapturedStdOutErr` which is appended to system out.

> 	// CapturedStdOutErr contains text printed to stdout/stderr (when running in parallel)
	// This is always empty when running in series or calling CurrentSpecReport()
	// It is used internally by Ginkgo's reporter
	CapturedStdOutErr string

chendave commented:
> trim ReportEntries with By Step as name (could go into this PR)

I trimmed the `ReportEntries` entirely in this pr.

> replace all ginkgo.By calls with normal framework.Logf calls (would also get rid of the annoying (?) escape sequences

What do you mean by "annoying (?) escape sequences"?

pohly commented:
> I don't have one atm

Good. Let's deal with it if we ever find that it's really non-empty.

> I trimmed the ReportEntries entirely in this pr.

But only for successful reports. My proposal is to trim them also for failed ones. The effect in most cases will be that we don't get the "open stdout" link in Spyglass because `system-out` will not be included anymore, only `system-err`. That's better for devs because it's less to things to check after a failure.

> What do you mean by "annoying (?) escape sequences"?

When open stderr, I get:
```
BeforeEach] [sig-node] Probing container
  test/e2e/framework/framework.go:186
ï¿½[1mSTEP:ï¿½[0m Creating a kubernetes client ï¿½[38;5;243m08/02/22 07:59:02.423ï¿½[0m
Aug  2 07:59:02.423: INFO: >>> kubeConfig: /root/.kube/kind-test-config
```

Note the escape sequences around "Creating a kubernetes client".

pohly commented:
@chendave gave me permission to update his branch while is out of the office for today. I've pushed a different commit which implements what I outlined above (`ginkgo --junit-report` and `-report-dir` create different JUnit reports, ginkgo.By entries get removed in the latter).

aojea commented:
> Note the escape sequences around "Creating a kubernetes client".

if you are able to sanitize the output I owe you :smile:

pohly commented:
It's simple: we just have to add `--no-color` to the ginkgo invocation in Prow.

pohly commented:
There's even 'GINKGO_NO_COLOR' in hack/ginkgo-e2e.sh for it. So why is that not being set by Prow jobs?

Should we perhaps flip the default?

pohly commented:
-> https://github.com/kubernetes/kubernetes/pull/111633

pohly commented:
Darn, not a single job failed! :sweat_smile:

pohly commented:
/test pull-kubernetes-e2e-kind-ipv6 pull-kubernetes-e2e-kind

pohly commented:
As the Gremlins are gone, I added a ghost failure to one of the storage tests. The result is rendered as expected (failures have stderr, no stdout):
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/111627/pull-kubernetes-e2e-kind/1554457702673944576

Tests without failures have only a minimal entry in the JUnit XML:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/111627/pull-kubernetes-e2e-kind/1554457702673944576/artifacts/junit_01.xml

onsi commented:
@chendave yes this is exactly the approach I was imagining around how to trim the report and generate your own custom-formatted JUnit report.

I appreciate that your downstream tooling can't handle all the information in the default JUnit file.  in part this is due to the (unfortunate) reality that the JUnit spec is fairly anemic and doesn't have a lot of rich structure.

With that said, I actually think you _do_ want to preserve all that additional information.  Consider, for a specific example, the flake we recently looked at in #111086 - some of the key pieces of information there were these 

```
03:45:47.846: INFO: error deleting namespace csi-mock-volumes-6183: timed out waiting for the condition
```

info logs that were attached to _passing_ specs.

More broadly, my sense is that a complex integration test suite like the e2e suite can be super hard to debug - especially when running in parallel and having the potential for two running specs to have an unforeseen impact on each other.  Having information in an accessible format that will allow you to find correlations and extract clues for where to look next is important.

The JUnit format isn't great for that - especially if the tooling built on top of it simply can't handle the amount of data being generated.  So I'd recommend the following:

- keep just the truncated JUnit file that you've generated.  The use-case for this file is to populate downstream tooling and allow users to quickly understand if a suite failed and to get information about just the failing specs.
- use ginkgo's `json-report` format to save an archive of the entire test run.  This will have a ton of detailed information.  The use-case here is to provide the possibility to dive deeper into the suite for debugging purposes.

Over time I could imagine building your own custom tooling/heuristics no top of the JSON format (e.g. something that detects whether or not there is an elevated rate of non-failing errors; or something that detects whether or not individual specs are taking longer to run than usual).

pohly commented:
> some of the key pieces of information there were these info logs that were attached to passing specs.

We still have that output in the https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/111627/pull-kubernetes-e2e-kind/1554471210773909504/build-log.txt. I suspect that such a full log will be more suitable for debugging such ugly cross-spec interactions because one can grep for things more easily and see what is running at which time.

We can certainly also add a full dump, but without tooling to consume it that's just going to cause costs for the gcp buckets.

chendave commented on test/e2e/framework/test_context.go:
```diff
@@ -330,14 +329,6 @@ func CreateGinkgoConfig() (types.SuiteConfig, types.ReporterConfig) {
-		// With Ginkgo v1, we used to write one file per parallel node. Now Ginkgo v2 automatically
-		// merges all results into a single file for us. The 01 suffix is kept in case that users
-		// expect files to be called "junit_<prefix><number>.xml".
-		reporterConfig.JUnitReport = path.Join(TestContext.ReportDir, "junit_"+TestContext.ReportPrefix+"01.xml")
```
By removing this, there will be only one junit file left, the trimmed one. This is what we discussed before and need to get the consensus on this.

aojea replied:
we only need one, right? the one that testgrid has to consume
+1
@pohly WDYT?

chendave replied:
yes, this is proposed by @pohly .  😄  let's collect some feedback on this. 

please note, we don't have a json file yet as what @onsi said.

chendave commented on test/e2e/e2e_test.go:
```diff
@@ -169,16 +171,66 @@ var _ = ginkgo.ReportAfterSuite("Kubernetes e2e suite report", func(report ginkg
+		// Remove report entries generated by ginkgo.By("doing
+		// something") because those are not useful (just have the
+		// start time) and cause Spyglass to show an additional "open
+		// stdout" button with a summary of the steps, which usually
```
we'd better to focus on issue here,  if we all agree that a stdout for a failure testcase are useless, we can address that in a separate pr, WDYT?

chendave replied:
Let's discuss this issue in another pr, I'd prefer to not touch this in this one.

aojea replied:
some test use them to log the steps of the test, I think we should keep them and merge, I think that was the previous behavior , but that is for another PR

chendave replied:
+1

pohly replied:
The text messages from `ginkgo.By` still shows up in in the output even when trimming report entries. That is independent from each other because `ginkgo.By` now does two things:
- log the output
- create a report entry

I can create a separate PR for trimming the report entries.

chendave committed:
e2e: trim junit report for Spyglass

Including the full information for successful tests makes the resulting XML
file too large for the 200GB limit in Spyglass when running large jobs (like
scale testing).

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>
Co-authored-by: Dave Chen <dave.chen@arm.com>

chendave commented:
> use ginkgo's json-report format to save an archive of the entire test run.

we add that in this pr: https://github.com/kubernetes/kubernetes/pull/111339, but seems like the json file is not created, https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/111339/pull-kubernetes-e2e-kind/1552131835801112576/artifacts/.

so, the major concern is that whether we need a default junit file, since we don't have a json report by default.

other than that, this one should be good to go.

/assign @BenTheElder @aojea @dims

pohly commented on test/e2e/e2e_test.go:
```diff
@@ -169,16 +171,58 @@ var _ = ginkgo.ReportAfterSuite("Kubernetes e2e suite report", func(report ginkg
+			specReport.CapturedStdOutErr = ""
+			// strip some details for tracing each steps executed by Ginkgo, this is used to build the "system-out"
+			// while "system-out" is not shown by Spyglass if the testcase is not failed.
+			specReport.ReportEntries = nil
```
Note that trimming this here is slightly too aggressive: if a test stores measurements during a successful run, then we probably should keep those entries because otherwise doing those measurements is questionable (where are they visible?).

I'm not sure whether this is a real problem and I will send a PR with my original code for trimming ReportEntries anyway as soon as this gets merged, so I don't mind having this go in as-is.

chendave replied:
> if a test stores measurements during a successful run, then we probably should keep those entries because otherwise doing those measurements is questionable

I wrote a simple test with measurements, the testcase passed without any issue, so I guess that entry added will only be used to generate a report, and while the testcase passed, those info is not shown by any downstream tools, so I think it should be fine.

+1 on more cleanup around this in a separate pr.

pohly replied:
Where could you see the result of the measurements?

chendave replied:
no, I can just see that testcase passed.

```
          <testcase name="[It] [sig-autoscaling] [Feature:ClusterSizeAutoscalingScaleUp] [Slow] Autoscaling Simple Check do nothing here should takes less than 1 minutes" classname="Kubernetes e2e suite" status="passed" time="0.063344167"></testcase>
```

aojea commented:
/approve

pohly.for lgtm

pohly commented:
/lgtm

pohly commented:
@dims, @BenTheElder: can someone set the milestone?

dims commented:
/milestone v1.25

Test Freeze is on the 10th August 2022. So this qualifies for the milestone as it has changes to test harness only
